api/datastore/sql/sql.go

```diff
@@ -689,6 +706,154 @@ func (ds *SQLStore) GetRoutesByApp(ctx context.Context, appID string, filter *mo
 	return res, nil
 }
 
+func (ds *SQLStore) PutFunc(ctx context.Context, fn *models.Func) (*models.Func, error) {
+	err := ds.Tx(func(tx *sqlx.Tx) error {
+		query := tx.Rebind(fmt.Sprintf(`%s WHERE name=?`, funcSelector))
+		row := tx.QueryRowxContext(ctx, query, fn.Name)
+
+		var dst models.Func
+		err := row.StructScan(&dst)
+		if err != nil && err != sql.ErrNoRows {
+			return err
+		} else if err == sql.ErrNoRows {
+			// insert it
+
+			fn.SetDefaults()
+			err = fn.Validate()
+			if err != nil {
+				return err
+			}
+
+			query = tx.Rebind(`INSERT INTO funcs (
+				id,
+				name,
+				image,
+				format,
+				memory,
+				cpus,
+				timeout,
+				idle_timeout,
+				config,
+				annotations,
+				created_at,
+				updated_at
+			)
+			VALUES (
+				:id,
+				:name,
+				:image,
+				:format,
+				:memory,
+				:cpus,
+				:timeout,
+				:idle_timeout,
+				:config,
+				:annotations,
+				:created_at,
+				:updated_at
+			);`)
+
+		} else {
+			// update it
+			dst.Update(fn)
+			err = dst.Validate()
+			if err != nil {
+				return err
+			}
+			fn = &dst // set for query & to return
+
+			query = tx.Rebind(`UPDATE funcs SET
```

Comment:
Does this mean update structs must contain all values?

Reply:
they don't have to as `dst.Update(fn)` will fill in all fields from the existing, updating any modified. one thing of note is that as it stands this keeps the strange config deletion failure which seems to get us a bug report every 6-8 weeks, we can change that (it would change app/route configs as well, so left for now), need to discuss UX as e.g. `fn apps config set myKey myValue` won't work as expected anymore in CLI

Reply:
Config deletion? It is a problem with complete update style apis, but fixing it can be an expense people tend not to like.

Reply:
sorry, I mean config key delete doesn't work if you simply update a config that omits a key. so if a user sends in a config `{"a":"v", "b":"v"}` when config `{"c":"v"}` exists, you get config `{"a":"v", "b":"v","c":"v"}` and it seems most of our users want the first config shown here to be the final config, not the one I've shown that simply adds new keys to existing. as it stands, to delete a key, users have to do `{"a":""}` (send an empty value), and thus far this has not been something intuited by users.

Reply:
Ah ok, the opposite implementation.

I think if we have CAS, we can do the first version, i.e. just replace the entire config rather than merge.